modAionImpl/src/org/aion/zero/impl/sync/TaskImportBlocks.java

```diff
@@ -64,204 +80,622 @@
+        if (importResult.isBest() || state.isOverRepeatThreshold()) {
+            state.setMode(NORMAL);
+        } else {
+            // in case we continue as FORWARD
+            state.setBase(lastBlock);
+        }
+
+        return state;
+    }
+
+    /**
+     * Imports blocks from storage as long as there are blocks to import.
+     *
+     * @return the total number of imported blocks from all iterations
+     */
+    private int importFromStorage(PeerState state, long first, long last) {
+        ImportResult importResult = ImportResult.NO_PARENT;
+        int imported = 0, batch;
+        long level = first;
+
+        while (level <= last) {
+            // get blocks stored for level
+            Map<ByteArrayWrapper, List<AionBlock>> levelFromDisk =
+                    chain.loadPendingBlocksAtLevel(level);
+
+            if (levelFromDisk.isEmpty()) {
+                // move on to next level
+                level++;
+                continue;
+            }
+
+            List<ByteArrayWrapper> importedQueues = new ArrayList<>(levelFromDisk.keySet());
+
+            for (Map.Entry<ByteArrayWrapper, List<AionBlock>> entry : levelFromDisk.entrySet()) {
+                // initialize batch counter
+                batch = 0;
+
+                List<AionBlock> batchFromDisk = entry.getValue();
+
+                if (log.isDebugEnabled()) {
+                    log.debug(
+                            "Loaded {} blocks from disk from level {} queue {} before filtering.",
+                            batchFromDisk.size(),
+                            entry.getKey(),
+                            level);
+                }
+
+                // filter already imported blocks
+                batchFromDisk = filterBatch(batchFromDisk, chain, importedBlockHashes);
+
+                if (batchFromDisk.size() > 0) {
+                    if (log.isDebugEnabled()) {
+                        log.debug(
+                                "{} {} left after filtering out imported blocks.",
+                                batchFromDisk.size(),
+                                (batchFromDisk.size() == 1 ? "block" : "blocks"));
+                    }
+                } else {
+                    if (log.isDebugEnabled()) {
+                        log.debug("No blocks left after filtering out imported blocks.");
+                    }
+                    // move on to next queue
+                    // this queue will be deleted from storage
+                    continue;
+                }
+
+                for (AionBlock b : batchFromDisk) {
+                    try {
+                        importResult = importBlock(b, "STORAGE", state);
+
+                        if (importResult.isStored()) {
+                            importedBlockHashes.put(ByteArrayWrapper.wrap(b.getHash()), true);
+
+                            batch++;
+
+                            if (last <= b.getNumber()) {
```

Comment:
has any change to import more blocks when last < b.getNumber()?

Reply:
you're right, `last` will be at most equal to the block number; correction made in b5bbd67